components/object_store/src/obkv/meta.rs

```diff
@@ -0,0 +1,373 @@
+    /// The version of object, we use the upload_id as version.
+    /// TODO: Since `upload_id` is used by multiple objects, it may become very
+    /// large. Should we assign a version number to each object to avoid
+    /// this issue?
+    #[serde(rename = "version")]
+    pub version: String,
+}
+
+impl ObkvObjectMeta {
+    #[inline]
+    pub fn decode(data: &[u8]) -> Result<Self> {
+        ensure!(
+            data[0] == HEADER,
+            InvalidHeader {
+                header: data[0],
+                expect: HEADER,
+            }
+        );
+        let json = str::from_utf8(&data[1..]).context(InvalidUtf8)?;
+        serde_json::from_str(json).context(InvalidJson { json })
+    }
+
+    #[inline]
+    pub fn encode(&self) -> Result<Vec<u8>> {
+        let size = self.estimate_size_of_json();
+        let mut encode_bytes = Vec::with_capacity(size + 1);
+        encode_bytes.push(HEADER);
+        serde_json::to_writer(&mut encode_bytes, self).context(EncodeJson)?;
+        Ok(encode_bytes)
+    }
+
+    /// Estimate the json string size of ObkvObjectMeta
+    #[inline]
+    pub fn estimate_size_of_json(&self) -> usize {
+        // {}
+        let mut size = 2;
+        // size of key name, `,`, `""` and `:`
+        size += (8 + 13 + 4 + 9 + 9 + 5 + 7) + 4 * 7;
+        size += self.location.len() + 2;
+        // last_modified
+        size += 8;
+        // size
+        size += 8;
+        // unique_id
+        if let Some(id) = &self.unique_id {
+            size += id.len() + 2;
+        } else {
+            size += 4;
+        }
+        // part_size
+        size += 8;
+        // parts
+        for part in &self.parts {
+            // part.len, `""`, `:`, and `,`
+            size += part.len() + 4;
+        }
+        //{}
+        size += 2;
+        // version
+        size += self.version.len();
+        size
+    }
+
+    /// Compute the convered parts based on given range parameter
+    pub fn compute_covered_parts(&self, range: Range<usize>) -> Result<ConveredParts> {
+        ensure!(
+            range.start <= range.end && range.end <= self.size,
+            InvalidRange {
+                start: range.start,
+                end: range.end,
+            }
+        );
+        let batch_size = self.part_size;
+        let start_index = range.start / batch_size;
+        let start_offset = range.start % batch_size;
+        let end_index = range.end / batch_size;
+        let end_offset = range.end % batch_size;
+
+        let mut convered_parts = Vec::with_capacity(end_index - start_index + 1);
+
+        for key in self.parts.iter().take(end_index + 1).skip(start_index) {
+            convered_parts.push(key.clone());
+        }
+
+        Ok(ConveredParts {
+            part_keys: convered_parts,
+            start_offset,
+            end_offset,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ConveredParts {
+    /// The table kv client
+    pub part_keys: Vec<String>,
```

Comment:
Maybe we can avoid memory allocation for part_keys:
```suggestion
pub struct ConveredParts<'a> {
    pub part_keys: &'a [String],
```

Reply:
Good idea.